src/System.Xml.XmlSerializer/src/System/Xml/Serialization/ReflectionXmlSerializationWriter.cs

```diff
@@ -0,0 +1,1112 @@
+                    continue;
+                return i;
+            }
+            return -1;
+        }
+
+        bool WriteDerivedTypes(StructMapping mapping, string n, string ns, object o, bool isNullable)
+        {
+            Type t = o.GetType();
+            for (StructMapping derived = mapping.DerivedMappings; derived != null; derived = derived.NextDerivedMapping)
+            {
+                if (t == derived.TypeDesc.Type)
+                {
+                    WriteStructMethod(derived, n, ns, o, isNullable, needType: true);
+                    return true;
+                }
+
+                if (WriteDerivedTypes(derived, n, ns, o, isNullable))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private void WritePrimitive(string method, string name, string ns, object defaultValue, object o, TypeMapping mapping, bool writeXsiType, bool isElement, bool isNullable)
+        {
+            TypeDesc typeDesc = mapping.TypeDesc;
+            bool hasDefault = defaultValue != null && !Globals.IsDBNullValue(defaultValue) && mapping.TypeDesc.HasDefaultSupport;
+            if (hasDefault)
+            {
+                if (mapping is EnumMapping)
+                {
+                    if (((EnumMapping)mapping).IsFlags)
+                    {
+                        var defaultEnumFlagValues = defaultValue.ToString().Split(null).Where((s) => !string.IsNullOrWhiteSpace(s));
+                        string defaultEnumFlagString = string.Join(", ", defaultEnumFlagValues);
+
+                        if (o.ToString() == defaultEnumFlagString)
+                            return;
+                    }
+                    else
+                    {
+                        if (o.ToString() == defaultValue.ToString())
+                            return;
+                    }
+                }
+                else
+                {
+                    if (IsDefaultValue(mapping, o, defaultValue, isNullable))
+                    {
+                        return;
+                    }
+                }
+            }
+
+            XmlQualifiedName xmlQualifiedName = null;
+            if (writeXsiType)
+            {
+                xmlQualifiedName = new XmlQualifiedName(mapping.TypeName, mapping.Namespace);
+            }
+
+            string stringValue = null;
+            bool hasValidStringValue = false;
+            if (mapping is EnumMapping)
+            {
+                stringValue = WriteEnumMethod((EnumMapping)mapping, o);
+                hasValidStringValue = true;
+            }
+            else
+            {
+                // WritePrimitiveValue(typeDesc, source, isElement);
+                hasValidStringValue = WritePrimitiveValue(typeDesc, o, isElement, out stringValue);
+            }
+
+            if (hasValidStringValue)
+            {
+                if (method == "WriteElementString")
+                {
```

Comment:
switch statement would be better. Even if it's a string, it's faster than cascaded if's